java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt

```diff
@@ -4702,6 +4708,30 @@ open class KotlinFileExtractor(
         }
     }
 
+    // todo: calculating the enclosing ref type could be done through this, instead of walking up the declaration parent chain
+    private val declarationStack = DeclarationStack()
+
+    private inner class DeclarationStack {
+        private val stack: Stack<IrDeclaration> = Stack()
+
+        fun push(item: IrDeclaration) = stack.push(item)
+
+        fun pop() = stack.pop()
+
+        fun isEmpty() = stack.isEmpty()
+
+        fun peek() = stack.peek()
+
+        fun peekAsDeclarationParent(): IrDeclarationParent {
+            val trapWriter = tw
+            if (isEmpty() && trapWriter is SourceFileTrapWriter) {
+                // If the current declaration is used as a parent, we might end up with an empty stack. In this case, the source file is the parent.
+                return trapWriter.irFile
+            }
+
+            return peek() as IrDeclarationParent
```

Comment:
Can this cast ever fail?

Reply:
Pushed https://github.com/github/codeql/pull/10239/commits/2728517c7f6d8deda20c8206ec27c38951662661 to handle the error.